cpp/include/cuspatial/experimental/array_view/multipoint_array.cuh

```diff
@@ -0,0 +1,145 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/experimental/geometry_collection/multipoint.cuh>
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/traits.hpp>
+
+namespace cuspatial {
+namespace array_view {
+
+using namespace cuspatial::geometry_collection;
+
+/**
+ * @brief Host-Device view object of a multipoint array
+ * @ingroup array_view
+ *
+ * Conforms to GeoArrow's specification of multipoint:
+ * https://github.com/geopandas/geo-arrow-spec/blob/main/format.md
+ *
+ * @tparam GeometryIterator iterator type for the offset array. Must meet
+ * the requirements of [LegacyRandomAccessIterator][LinkLRAI].
+ * @tparam VecIterator iterator type for the point array. Must meet
+ * the requirements of [LegacyRandomAccessIterator][LinkLRAI].
+ *
+ * @note Though this object is host/device compatible,
+ * The underlying iterator should be device accessible if used in device kernel.
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename GeometryIterator, typename VecIterator>
+class multipoint_array {
+ public:
+  using geometry_it_t = GeometryIterator;
+  using point_it_t    = VecIterator;
+  using point_t       = iterator_value_type<point_it_t>;
+  using element_t     = iterator_vec_base_type<point_it_t>;
+
+  /**
+   * @brief Construct a new multipoint array object
+   */
+  multipoint_array(GeometryIterator geometry_begin,
+                   GeometryIterator geometry_end,
+                   VecIterator points_begin,
+                   VecIterator points_end);
+
+  /**
+   * @brief Returns the number of multipoints in the array.
+   */
+  auto size();
+
+  /**
+   * @brief Returns the iterator to the start of the multipoint array.
+   */
+  auto multipoint_begin();
+
+  /**
+   * @brief Returns the iterator to the end of the multipoint array.
+   */
+  auto multipoint_end();
+
+  /**
+   * @brief Returns the iterator to the start of the underlying point array.
+   */
+  auto point_begin();
+
+  /**
+   * @brief Returns the iterator to the end of the underlying point array.
+   */
+  auto point_end();
+
+  /**
+   * @brief Returns the `idx`th multipoint in the array.
+   *
+   * @tparam IndexType type of the index
+   * @param idx the index to the multipoint
+   * @return a multipoint object
+   */
+  template <typename IndexType>
+  CUSPATIAL_HOST_DEVICE auto element(IndexType idx);
+
+ protected:
+  /// Iterator to the start of the index array of start positions to each multipoint.
+  GeometryIterator _geometry_begin;
+  /// Iterator to the past-the-end of the index array of start positions to each multipoint.
+  GeometryIterator _geometry_end;
+  /// Iterator to the start of the point array.
+  VecIterator _points_begin;
+  /// Iterator to the past-the-end position of the point array.
+  VecIterator _points_end;
```

Comment:
I believe they are equivalent. Because of `LRAI` requirement on these iterators. You can always derive `begin_iterator` and `size` from these iterators, and vice versa.

One exception is that if you use `counting_iterator` on the geometry offset arrays, some iterator traits of the span might be ill-defined (e.g. `reference`).

With that in mind, I'm willing to adapt the design to look more closely to `std::span`.

Reply:
A more fully fleshed out example is shown here: https://godbolt.org/z/vx9Y897bc

Reply:
Decided that `span` isn't the appropriate model for this structure. `ranges` is.